src/commons/crypto/signing/signers/pkcs11/internal.rs

```diff
@@ -0,0 +1,981 @@
+use std::{
+    convert::TryInto,
+    path::Path,
+    sync::{Arc, RwLock, RwLockReadGuard},
+    time::Duration,
+};
+
+use backoff::ExponentialBackoff;
+
+use bcder::encode::{PrimitiveContent, Values};
+use bytes::Bytes;
+use pkcs11::errors::Error as Pkcs11Error;
+use pkcs11::types::*;
+use rpki::repository::crypto::{
+    signer::KeyError, KeyIdentifier, PublicKey, PublicKeyFormat, Signature, SignatureAlgorithm,
+};
+
+use crate::commons::{
+    api::Handle,
+    crypto::{
+        dispatch::signerinfo::SignerMapper,
+        signers::{
+            pkcs11::{
+                context::{Pkcs11Context, ThreadSafePkcs11Context},
+                session::Pkcs11Session,
+            },
+            probe::{ProbeError, ProbeStatus, StatefulProbe},
+            util,
+        },
+        SignerError,
+    },
+};
+
+//------------ Types and constants ------------------------------------------------------------------------------------
+
+/// The time to wait between an initial and subsequent attempt at sending a request to the PKCS#11 server.
+const RETRY_REQ_AFTER: Duration = Duration::from_secs(2);
+
+/// How much longer should we wait from one request attempt to the next compared to the previous wait?
+const RETRY_REQ_AFTER_MULTIPLIER: f64 = 1.5;
+
+/// The maximum amount of time to keep retrying a failed request.
+const RETRY_REQ_UNTIL_MAX: Duration = Duration::from_secs(30);
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum LoginMode {
+    // The token can do cryptographic operations such as signing without requiring C_Login to be called first, and so a
+    // user pin is also not required.
+    LoginNotRequired,
+
+    // The token requires that C_Login be called prior to performing any cryptographic operations such as signing. A
+    // correct user pin may be needed for the login to succeed.
+    LoginRequired,
+}
+
+// Placeholder struct
+#[derive(Clone, Debug)]
+struct ConnectionSettings {
+    context: ThreadSafePkcs11Context,
+
+    // For some PKCS#11 libraries it is easy, or only possible, to connect by slot ID (rather than slot label). If
```

Comment:
If one MUST be some, then an enum would be more intuitive to me here.

Reply:
Good point. Done. I'll need to merge it down through the dependent PRs as well. The commit that implemented the change also refactored `fn probe_server()` - sorry about that, it makes the commit harder to review but I had to touch it and foundcode unreasonably hard to follow. Perhaps it should even be factored out entirely into a separate struct or module ...